Trying to take a look at this, but I'm seeing compilation issues:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project opencast-speech-to-text-impl: Compilation failure: Compilation failure: 
[ERROR] /home/greg/opencast/upstream/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:[62,8] org.opencastproject.speechtotext.impl.engine.WhisperCppEngine is not abstract and does not override abstract method generateSubtitlesFile(java.io.File,java.io.File,java.lang.String,java.lang.Boolean) in org.opencastproject.speechtotext.api.SpeechToTextEngine
[ERROR] /home/greg/opencast/upstream/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:[260,30] generateSubtitlesFile(java.io.File,java.io.File,java.lang.String,java.lang.Boolean) in org.opencastproject.speechtotext.impl.engine.WhisperCppEngine cannot implement generateSubtitlesFile(java.io.File,java.io.File,java.lang.String,java.lang.Boolean) in org.opencastproject.speechtotext.api.SpeechToTextEngine
[ERROR]   return type java.util.Map<java.lang.String,java.lang.Object> is not compatible with org.opencastproject.speechtotext.api.SpeechToTextEngine.Result
[ERROR] /home/greg/opencast/upstream/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:[259,3] method does not override or implement a method from a supertype

```

--- a/docs/guides/admin/docs/configuration/transcription.configuration/whispercpp.md
+++ b/docs/guides/admin/docs/configuration/transcription.configuration/whispercpp.md
@@ -1,0 +1,71 @@
+WhisperC++ Transcription Engine
+===============================
+
+Overview
+--------
+
+Opencast can take advantage of [Open AI's Whisper Project](https://github.com/openai/whisper) to generate automatic
+transcriptions on premise through [SpeechToText WoH](../../workflowoperationhandlers/speechtotext-woh.md).
+
+WhisperC++ is a C/C++ implementation of OpenAI's Whisper automatic speech recognition (ASR) model.
+
+Advantages
+----------
+- Transcription on more than 80 languages
+- Translation to English
+- Automatic language detection
+- Faster processing than whisper when using CPUs
+- Run locally, no data sent to third parties.
+
+Enable WhisperC++ engine
+------------------------
+
+To enable WhisperC++ as for the `SpeechToText` WoH, follow these steps.
+
+1. Install whispercpp binary and language models following the instruction on
+[whisper.cpp](https://github.com/ggerganov/whisper.cpp) repository or build and use your own RPMs for RHEL-based
+distributions from [whispercpp-rpmbuild](https://github.com/elearning-univie/whispercpp-rpmbuild).
+2. Enable whispercpp engine and set job load in `org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.cfg`.
+3. Set the target model to use in `org.opencastproject.speechtotext.impl.engine.WhisperCppEngine`.
+4. WhisperC++ processes only PCM16 (wav) audio files. Therefore you probably have to add an `encode`-operation before
+running `speechtotext` in your workflow and an encoder profile:
+
+```
+  - id: encode
+    description: "Extract audio for processing with whispercpp"
+    configurations:
+      - source-flavor: "*/themed"
+      - target-flavor: "*/audio+stt"
+      - encoding-profile: audio-whisper
+
+  - id: speechtotext
+    description: "Generates subtitles with whispercpp"
+    configurations:
+      - source-flavor: "*/audio+stt"
+      - target-flavor: captions/vtt+auto-#{lang}
+      - target-element: attachment
+      - target-tags: archive,subtitle,engage-download
+      # - translate: true
+```
+
+```
+# Audio-only encoding format used for whispercpp
+profile.audio-whisper.name = whispercpp wav
+profile.audio-whisper.input = audio
+profile.audio-whisper.output = audio
+profile.audio-whisper.suffix = -stt.wav
+profile.audio-whisper.ffmpeg.command = -i #{in.video.path} -vn -ar 16000 -ac 1 -c:a pcm_s16le #{out.dir}/#{out.name}#{out.suffix}
+profile.audio-whisper.jobload = 1.0
+```
+
+
+Additional Notes
+----------------
+
+- Considerations about model usage and job load configuration in the [Additional notes](whisper.md#additional-notes)
+  from the `Whisper Transcription Engine` are also valid for this engine.
+- WhisperC++ can be run on CPU only or (partially) on dedicated hardware (Nvidia GPU, Apple Core ML) - if whispercpp
+  binary/library is built with appropriate options/modules.
+- Language models are based on Whisper and are transformed to a custom ggml format. You can download already converted
+  models or build and tweak your own models. For more informations visit
+  [models/README.md](https://github.com/ggerganov/whisper.cpp/blob/master/models/README.md).

--- a/docs/guides/admin/docs/workflowoperationhandlers/speechtotext-woh.md
+++ b/docs/guides/admin/docs/workflowoperationhandlers/speechtotext-woh.md
@@ -6,8 +6,8 @@ ID: `speechtotext`
 Description
 -----------
 
-The speech to text operation can be used to generate subtitles for Videos or Audio files. Currently, there are three STT
-engines available, [Whisper](../configuration/transcription.configuration/whisper.md),
+The speech to text operation can be used to generate subtitles from video or audio files. Currently, there are three
+STT engines available, [Whisper](../configuration/transcription.configuration/whisper.md),
 [WhisperC++](../configuration/transcription.configuration/whispercpp.md) and
 [Vosk](../configuration/transcription.configuration/vosk.md).
 The subtitles file format ist WebVTT.

--- a/etc/org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.cfg
+++ b/etc/org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.cfg
@@ -3,5 +3,5 @@
 #job.load.speechtotext=0.8
 
 # Select STT engine.
-# Available engines: vosk, whisper | Default: vosk
-SpeechToTextEngine.target=(enginetype=whisper)
+# Available engines: vosk, whisper, whispercpp | Default: vosk
+#SpeechToTextEngine.target=(enginetype=vosk)

--- a/etc/org.opencastproject.speechtotext.impl.engine.WhisperCppEngine.cfg
+++ b/etc/org.opencastproject.speechtotext.impl.engine.WhisperCppEngine.cfg
@@ -1,0 +1,9 @@
+# Configuration for setting a custom path to the Whisper command line tool
+# Default: whisper
+#whispercpp.root.path=whispercpp
+
+# Absolute path to language model
+# RPMs install models to /usr/share/ggml/ggml-<model>.bin
+# Available models: tiny, base, small, medium, large.
+# Default: /usr/share/ggml/ggml-base.bin
+#whispercpp.model=/usr/share/ggml/ggml-base.bin

--- a/etc/org.opencastproject.speechtotext.impl.engine.WhisperCppEngine.cfg
+++ b/etc/org.opencastproject.speechtotext.impl.engine.WhisperCppEngine.cfg
@@ -7,3 +7,12 @@
 # Available models: tiny, base, small, medium, large.
 # Default: /usr/share/ggml/ggml-base.bin
 #whispercpp.model=/usr/share/ggml/ggml-base.bin
+
+
+## Optional settings to configure the underlying whispercpp engine
+
+# Beam size for beam search
+#whispercpp.beam.size=-1
+
+# Maximum segment length in characters
+#whispercpp.max.length=0

--- a/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java
+++ b/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java
@@ -1,0 +1,197 @@
+/**
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ *
+ * The Apereo Foundation licenses this file to you under the Educational
+ * Community License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License
+ * at:
+ *
+ *   http://opensource.org/licenses/ecl2.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.opencastproject.speechtotext.impl.engine;
+
+import org.opencastproject.speechtotext.api.SpeechToTextEngine;
+import org.opencastproject.speechtotext.api.SpeechToTextEngineException;
+import org.opencastproject.util.IoSupport;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Modified;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileReader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** WhisperC++ implementation of the Speech-to-text engine interface. */
+@Component(
+    property = {
+        "service.description=WhisperC++ implementation of the SpeechToTextEngine interface",
+        "enginetype=whispercpp"
+    }
+)
+
+public class WhisperCppEngine implements SpeechToTextEngine {
+
+  private static final Logger logger = LoggerFactory.getLogger(WhisperCppEngine.class);
+
+  /** Name of the engine. */
+  private static final String engineName = "WhisperC++";
+
+  /** Config key for setting the path to WhisperC++. */
+  private static final String WHISPERCPP_EXECUTABLE_PATH_CONFIG_KEY = "whispercpp.root.path";
+
+  /** Default path to WhisperC++. */
+  public static final String WHISPERCPP_EXECUTABLE_DEFAULT_PATH = "whispercpp";
+
+  /** Currently used path of the WhisperC++ installation. */
+  private String whispercppExecutable = WHISPERCPP_EXECUTABLE_DEFAULT_PATH;
+
+  /** Config key for setting whispercpp model */
+  private static final String WHISPERCPP_MODEL_CONFIG_KEY = "whispercpp.model";
+
+  /** Default whispercpp model */
+  public static final String WHISPERCPP_MODEL_DEFAULT = "/usr/share/ggml/ggml-base.bin";
+
+  /** Currently used whispercpp model */
+  private String whispercppModel = WHISPERCPP_MODEL_DEFAULT;
+
+
+  @Override
+  public String getEngineName() {
+    return engineName;
+  }
+
+  @Activate
+  @Modified
+  public void activate(ComponentContext cc) {
+    logger.debug("Activated/Modified WhisperC++ engine service class");
+    whispercppExecutable = StringUtils.defaultIfBlank(
+        (String) cc.getProperties().get(WHISPERCPP_EXECUTABLE_PATH_CONFIG_KEY), WHISPERCPP_EXECUTABLE_DEFAULT_PATH);
+    logger.debug("Set WhisperC++ path to {}", whispercppExecutable);
+
+    whispercppModel = StringUtils.defaultIfBlank(
+        (String) cc.getProperties().get(WHISPERCPP_MODEL_CONFIG_KEY), WHISPERCPP_MODEL_DEFAULT);
+    logger.debug("WhisperC++ Language model set to {}", whispercppModel);
+
+    logger.debug("Finished activating/updating speech-to-text service");
+  }
+
+  //TODO: Add method for language detection
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.opencastproject.speechtotext.api.SpeechToTextEngine#generateSubtitlesFile(File, File, String, Boolean)
+   */
+
+  @Override
+  public Map<String, Object> generateSubtitlesFile(File mediaFile,
+      File preparedOutputFile, String language, Boolean translate)
+          throws SpeechToTextEngineException {
+
+    String[] baseCommands = { whispercppExecutable,
+        mediaFile.getAbsolutePath(),
+        "--model", whispercppModel,
+        "-ovtt",
+        "-oj",
+        "-bs 5",
+        "--output-file", preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", "")};
+
+    List<String> command = new ArrayList<>(Arrays.asList(baseCommands));
+
+    if (translate) {
+      command.add("--translate");
+      logger.info("Translation enabled");
+      language = "en";
+    }
+
+    if (!language.isBlank() && !translate) {
+      logger.info("Using language {} from workflows", language);
+      command.add("--language");
+      command.add(language);
+    }
+
+    // TODO: Remove as this is used only for debugging
+    logger.info("Executing WhisperC++'s transcription command: {}", command);
+    logger.info("preparedOutputFile: {}", preparedOutputFile);
+    logger.info("mediaFile: {}", mediaFile.getAbsolutePath());
+    logger.info("preparedOutputFileNoExt: {}", preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", ""));
+
+    Process process = null;
+
+    try {
+      ProcessBuilder processBuilder = new ProcessBuilder(command);
+      processBuilder.redirectErrorStream(true);
+      process = processBuilder.start();
+
+
+      // wait until the task is finished
+      int exitCode = process.waitFor();
+      logger.info("WhisperC++ process finished with exit code {}",exitCode);
+
+      if (exitCode != 0) {
+        var error = "";
+        try (var errorStream = process.getInputStream()) {
+          error = "\n Output:\n" + IOUtils.toString(errorStream, StandardCharsets.UTF_8);
+        }
+        throw new SpeechToTextEngineException(
+            String.format("WhisperC++ exited abnormally with status %d (command: %s)%s", exitCode, command, error));
+      }
+
+      if (!preparedOutputFile.isFile()) {
+        throw new SpeechToTextEngineException("WhisperC++ produced no output");
+      }
+      logger.info("Subtitles file generated successfully: {}", preparedOutputFile);
+    } catch (Exception e) {
+      logger.info("Transcription failed closing WhisperC++ transcription process for: {}", mediaFile);
+      throw new SpeechToTextEngineException(e);
+    } finally {
+      IoSupport.closeQuietly(process);
+    }
+
+    // Detect language if not set
+    if (language.isBlank()) {
+      JSONParser jsonParser = new JSONParser();
+      try {
+        FileReader reader = new FileReader(preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", "")
+            + ".json");
+        Object obj = jsonParser.parse(reader);
+        JSONObject jsonObject = (JSONObject) obj;
+        language = (String) jsonObject.get("language");
+        logger.info("Language detected by WhisperC++: {}", language);
+      } catch (Exception e) {
+        logger.info("Error reading WhisperC++ JSON file for: {}", mediaFile);
+        throw new SpeechToTextEngineException(e);
+      }
+    }
+
+    Map<String,Object> returnValues = new HashMap<>();
+    returnValues.put("subFile",preparedOutputFile);
+    returnValues.put("language",language);
+
+    return returnValues; // Subtitles data
+  }
+}

--- a/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java
+++ b/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java
@@ -30,6 +30,8 @@ import org.apache.commons.io.IOUtils;
 import org.apache.commons.lang3.StringUtils;
 import org.json.simple.JSONObject;
 import org.json.simple.parser.JSONParser;
+import org.opencastproject.util.OsgiUtil;
+import org.opencastproject.util.data.Option;
 import org.osgi.service.component.ComponentContext;
 import org.osgi.service.component.annotations.Activate;
 import org.osgi.service.component.annotations.Component;
@@ -79,6 +81,18 @@ public class WhisperCppEngine implements SpeechToTextEngine {
   /** Currently used whispercpp model */
   private String whispercppModel = WHISPERCPP_MODEL_DEFAULT;
 
+  /** Config key for setting whispercpp beam size */
+  private static final String WHISPERCPP_BEAM_SIZE_CONFIG_KEY = "whispercpp.beam.size";
+
+  /** Currently used whispercpp beam size */
+  private Option<Integer> whispercppBeamSize;
+
+  /** Config key for setting whispercpp maximum segment length */
+  private static final String WHISPERCPP_MAX_LENGTH_CONFIG_KEY = "whispercpp.max.length";
+
+  /** Currently used whispercpp maximum segment length */
+  private Option<Integer> whispercppMaxLength;
+
 
   @Override
   public String getEngineName() {
@@ -96,6 +110,16 @@ public class WhisperCppEngine implements SpeechToTextEngine {
     whispercppModel = StringUtils.defaultIfBlank(
         (String) cc.getProperties().get(WHISPERCPP_MODEL_CONFIG_KEY), WHISPERCPP_MODEL_DEFAULT);
     logger.debug("WhisperC++ Language model set to {}", whispercppModel);
+
+    whispercppBeamSize = OsgiUtil.getOptCfgAsInt(cc.getProperties(), WHISPERCPP_BEAM_SIZE_CONFIG_KEY);
+    if (whispercppBeamSize.isSome()) {
+      logger.debug("WhisperC++ beam size set to {}", whispercppBeamSize);
+    }
+
+    whispercppMaxLength = OsgiUtil.getOptCfgAsInt(cc.getProperties(), WHISPERCPP_MAX_LENGTH_CONFIG_KEY);
+    if (whispercppMaxLength.isSome()) {
+      logger.debug("WhisperC++ maximum segment length set to {}", whispercppMaxLength);
+    }
 
     logger.debug("Finished activating/updating speech-to-text service");
   }
@@ -122,10 +146,19 @@ public class WhisperCppEngine implements SpeechToTextEngine {
         "--model", whispercppModel,
         "-ovtt",
         "-oj",
-        "-bs", "5",
         "--output-file", preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", "")};
 
     List<String> command = new ArrayList<>(Arrays.asList(baseCommands));
+
+    if (whispercppBeamSize.isSome()) {
+      command.add("-bs");
+      command.add(Integer.toString(whispercppBeamSize.get()));
+    }
+
+    if (whispercppMaxLength.isSome()) {
+      command.add("-ml");
+      command.add(Integer.toString(whispercppMaxLength.get()));
+    }
 
     String subtitleLanguage;
 

